src/sidechaindb.cpp

```diff
@@ -308,179 +360,238 @@ bool SidechainDB::Update(int nHeight, const uint256& hashBlock, const std::vecto
 
-        // Move on to this sidechain's next WT^
-        if (script[i] == SCOP_WT_DELIM) {
-            nWTIndex++;
-            continue;
+        // Get MT hash from script
+        CScript::const_iterator phash = scriptPubKey.begin() + 6;
+        opcodetype opcode;
+        std::vector<unsigned char> vch;
+        if (scriptPubKey.GetOp(phash, opcode, vch) && vch.size() == 32) {
+            // Try and sync
+            uint256 hashMerkleRoot = uint256(vch);
+            bool fUpdated = UpdateSCDBMatchMT(hashMerkleRoot);
+            // TODO handle !fUpdated
         }
+    }
 
-        // Move on to the next sidechain
-        if (script[i] == SCOP_SC_DELIM) {
-            nWTIndex = 0;
-            nSidechainIndex++;
-            continue;
-        }
+    // Update hashBLockLastSeen
+    hashBlockLastSeen = hashBlock;
 
-        // Check for valid vote type
-        const unsigned char& vote = script[i];
-        if (vote != SCOP_REJECT && vote != SCOP_VERIFY && vote != SCOP_IGNORE)
-            continue;
+    return true;
+}
 
-        if (nSidechainIndex > vState.size())
-            return false;
-        if (nWTIndex > vState[nSidechainIndex].size())
+bool SidechainDB::UpdateSCDBIndex(const std::vector<SidechainWTJoinState>& vNewScores)
+{
+    // First check that sidechain numbers are valid
+    for (const SidechainWTJoinState& s : vNewScores) {
+        if (!SidechainNumberValid(s.nSidechain))
             return false;
+    }
 
-        const SidechainWTJoinState& old = vState[nSidechainIndex][nWTIndex];
-
-        uint16_t nBlocksLeft = old.nBlocksLeft;
-        if (nBlocksLeft > 0)
-            nBlocksLeft--;
+    // Decrement nBlocksLeft of existing WT^(s)
+    for (const Sidechain& s : ValidSidechains) {
+        SCDBIndex& index = SCDB[s.nSidechain];
+        for (SidechainWTJoinState wt : index.members) {
+            // wt is a copy
+            wt.nBlocksLeft--;
+            index.InsertMember(wt);
+        }
+    }
 
-        uint16_t nWorkScore = old.nWorkScore;
-        if (vote == SCOP_REJECT) {
-            if (nWorkScore > 0)
-                nWorkScore--;
+    // Apply new work scores
+    for (const SidechainWTJoinState& s : vNewScores) {
+        SCDBIndex& index = SCDB[s.nSidechain];
+        SidechainWTJoinState wt;
+        if (index.GetMember(s.wtxid, wt)) {
+            // Update an existing WT^
+            // Check that new work score is valid
+            if ((wt.nWorkScore == s.nWorkScore) ||
+                    (s.nWorkScore == (wt.nWorkScore + 1)) ||
+                    (s.nWorkScore == (wt.nWorkScore - 1)))
+            {
+                index.InsertMember(s);
+            }
         }
         else
-        if (vote == SCOP_VERIFY) {
-            nWorkScore++;
+        if (!index.IsFull()) {
+            // Add a new WT^
+            if (!s.nWorkScore == 0)
+                continue;
+            if (s.nBlocksLeft != ValidSidechains[s.nSidechain].GetTau())
```

Comment:
Considering Paul wants the sidechain WT^ limit removed, we don't need to do the ```if (!index.IsFull()) ``` check anymore and can certainly simplify this code as you have suggested. Good idea :+1: